File: pkg-r/tests/testthat/test-species.R
test_that(".ensembl_id_prefix extracts the species code", {
  expect_identical(.ensembl_id_prefix("ENSG00000139618"), "")
  expect_identical(.ensembl_id_prefix("ENSMUSG00000059552"), "MUS")
  expect_identical(.ensembl_id_prefix("not-an-ensembl-id"), NA_character_)
})

test_that("a human id matches the human species", {
  res <- check_id(
    "ENSG00000139618",
    source_db = "ensembl",
    species = "homo_sapiens"
  )
  expect_true(res$valid)
  expect_identical(res$normalized, "ENSG00000139618")
})

test_that("a mouse id is invalid when human is requested, with no suggestion", {
  res <- check_id(
    "ENSMUSG00000059552",
    source_db = "ensembl",
    species = "homo_sapiens"
  )
  expect_false(res$valid)
  expect_true(is.na(res$normalized))
  expect_true(is.na(res$suggestion))
})

test_that("a taxon id matches the same species by name", {
  res <- check_id("ENSG00000139618", source_db = "ensembl", species = 9606)
  expect_true(res$valid)
  expect_identical(res$normalized, "ENSG00000139618")
})

test_that("an unknown species is not checked", {
  res <- check_id(
    "ENSG00000139618",
    source_db = "ensembl",
    species = "platypus"
  )
  expect_true(res$valid)
  expect_identical(res$normalized, "ENSG00000139618")
})

test_that("a species check applies to suggestions", {
  matched <- check_id(
    "ensmusg00000059552",
    source_db = "ensembl",
    species = "mus_musculus"
  )
  expect_false(matched$valid)
  expect_identical(matched$suggestion, "ENSMUSG00000059552")

  mismatched <- check_id(
    "ensmusg00000059552",
    source_db = "ensembl",
    species = "homo_sapiens"
  )
  expect_false(mismatched$valid)
  expect_true(is.na(mismatched$suggestion))
})

test_that("a non-species-aware source ignores species", {
  res <- check_id(
    "MONDO:0005148",
    source_db = "mondo",
    species = "homo_sapiens"
  )
  expect_true(res$valid)
  expect_identical(res$normalized, "MONDO:0005148")
})

File: pkg-py/tests/test_species.py
"""Offline, species-aware validation for Ensembl in pattern mode."""

import biobouncer
from biobouncer._pattern import _ensembl_id_prefix

HUMAN = "ENSG00000139618"
MOUSE = "ENSMUSG00000059552"
RAT = "ENSRNOG00000010756"


def _check(ident, species, source_db="ensembl"):
    return biobouncer.check_id(
        ident, source_db=source_db, how="pattern", species=species
    )[0]


def test_human_id_with_human_species_is_valid():
    res = _check(HUMAN, "homo_sapiens")
    assert res.valid is True
    assert res.normalized == HUMAN
    assert res.suggestion is None


def test_mouse_id_with_human_species_is_invalid():
    res = _check(MOUSE, "homo_sapiens")
    assert res.valid is False
    assert res.normalized is None
    assert res.suggestion is None


def test_human_id_with_taxon_species_is_valid():
    res = _check(HUMAN, 9606)
    assert res.valid is True
    assert res.normalized == HUMAN


def test_unknown_species_is_lenient():
    res = _check(HUMAN, "platypus")
    assert res.valid is True
    assert res.normalized == HUMAN


def test_lowercase_mouse_id_suggests_when_species_matches():
    res = _check(MOUSE.lower(), "mus_musculus")
    assert res.valid is False
    assert res.normalized is None
    assert res.suggestion == MOUSE


def test_lowercase_mouse_id_no_suggestion_when_species_mismatches():
    res = _check(MOUSE.lower(), "homo_sapiens")
    assert res.valid is False
    assert res.normalized is None
    assert res.suggestion is None


def test_rat_id_with_human_species_is_invalid():
    res = _check(RAT, "homo_sapiens")
    assert res.valid is False
    assert res.normalized is None
    assert res.suggestion is None


def test_lowercase_rat_id_suggests_when_species_matches():
    res = _check(RAT.lower(), "rattus_norvegicus")
    assert res.valid is False
    assert res.suggestion == RAT


def test_non_species_aware_source_ignores_species():
    res = _check("MONDO:0005148", "homo_sapiens", source_db="mondo")
    assert res.valid is True
    assert res.normalized == "MONDO:0005148"


def test_ensembl_id_prefix_extraction():
    assert _ensembl_id_prefix("ENSG00000139618") == ""
    assert _ensembl_id_prefix("ENSMUSG00000059552") == "MUS"
    assert _ensembl_id_prefix("ENSRNOG00000010756") == "RNO"
    assert _ensembl_id_prefix("MONDO:0005148") is None
